File: Python3/1488. Avoid Flood in The City.py
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        no_rain_count = []
        ans = [1] * len(rains)
        full = {}

        for i, r in enumerate(rains):
            if r == 0:
                no_rain_count.append(i)
                continue
            
            ans[i] = -1
            if r in full:
                p = bisect_left(no_rain_count, full[r])
                if p == len(no_rain_count):
                    return []
                
                ans[no_rain_count[p]] = r
                no_rain_count.pop(p)
            full[r] = i
                
        return ans

File: Java/1488. Avoid Flood in The City.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        int[] ans = new int[rains.length];
        Arrays.fill(ans, 1);;
        TreeSet<Integer> no_rain_days = new TreeSet<>();
        Map<Integer, Integer> full = new HashMap<>();
        for (int i = 0; i < rains.length; i++) {
            int lake = rains[i];
            if (lake == 0) {
                no_rain_days.add(i);
                continue;
            } 

            ans[i] = -1;
            if (full.containsKey(lake)) {
                Integer p = no_rain_days.ceiling(full.get(lake));
                if (p == null) {
                    return new int[0];
                }
                ans[p] = lake;
                no_rain_days.remove(p);
            }
            full.put(lake, i);
        }

        return ans;
    }
}